### ISSUE TITLE
Akka HTTP server graceful shutdown

### ISSUE THREAD

**Issue by RichardBradley** [akka/akka#16894]
We'd like to be able to do a "graceful shut down" of an Akka HTTP server, i.e. to unbind from the TCP port so as to stop accepting new requests, but to allow the currently open requests to complete (for up to a configurable timeout).

See [https://groups.google.com/forum/#!msg/spray-user/Sfh8x5yWibU/I1gDhwoULYsJ] for the equivalent feature in Spray 1.3.1

(Sorry if there is already a ticket for this, but I couldn't find one. I hope it's OK to create a new issue to track it here.)

**Comment by RichardBradley** [comment]
Sorry, duplicate of https://github.com/akka/akka/issues/15945

**Comment by jrudolph** [comment]
Reopened. Let's track the improved unbinding support in this ticket and use #15945 to document the current behavior.

**Comment by FeiWongReed** [comment]
I second this.

**Comment by ktoso** [comment]
This is actually the current behaviour, we'll want to document even better it which is the ticket: https://github.com/akka/akka/issues/20177

The "kill connections" one is shutdownNow which we do not have https://github.com/akka/akka/issues/20219

**Comment by jamesmulcahy** [comment]
@ktoso Are you sure?  When I call unbind() and wait on the future, it completes while I still have outstanding API requests.

Is unbind() not the correct method to call to take advantage of this behavior?

**Comment by reid-spencer** [comment]
@ktoso - Graceful shutdown is *not* the current behavior. The current behavior (2.4.8) is this:
```
java.util.NoSuchElementException: Binding was unbound before it was completely finished
	at akka.stream.impl.io.ConnectionSourceStage$$anon$1.postStop(TcpStages.scala:139)
	at akka.stream.impl.fusing.GraphInterpreter.finalizeStage(GraphInterpreter.scala:683)
	at akka.stream.impl.fusing.GraphInterpreter.finish(GraphInterpreter.scala:493)
	at akka.stream.impl.fusing.GraphInterpreterShell.tryAbort(ActorGraphInterpreter.scala:509)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$postStop$1.apply(ActorGraphInterpreter.scala:641)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$postStop$1.apply(ActorGraphInterpreter.scala:641)
	at scala.collection.immutable.Set$Set1.foreach(Set.scala:94)
	at akka.stream.impl.fusing.ActorGraphInterpreter.postStop(ActorGraphInterpreter.scala:641)
	at akka.actor.Actor$class.aroundPostStop(Actor.scala:494)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundPostStop(ActorGraphInterpreter.scala:529)
	at akka.actor.dungeon.FaultHandling$class.akka$actor$dungeon$FaultHandling$$finishTerminate(FaultHandling.scala:210)
	at akka.actor.dungeon.FaultHandling$class.terminate(FaultHandling.scala:172)
	at akka.actor.ActorCell.terminate(ActorCell.scala:374)
	at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:467)
	at akka.actor.ActorCell.systemInvoke(ActorCell.scala:483)
	at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:282)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:260)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```
That happens when we unbind the ServerBinding while requests are still in progress. Our system needs to have the ability to pause our REST Server and resume it, potentially multiple times per run of a JVM instance. So, if you need a test case, please let me know and I will provide. If you believe that graceful shutdown *is* implemented, I'd like to know how, please. Using the unbind method is what I gleaned from the docs and the example code I could find.

**Comment by drewhk** [comment]
Unbind means exactly that: unbind the port. That does not mean that
*already accepted* connections are stopped or that it waits on them. When
the unbind future completes, it only means that no more connections will be
accepted.

-Endre

**Comment by TJSomething** [comment]
I was looking for this, and my workaround was:

```scala
val serverSource = Http().bind(host, port)
val serverSettings = ServerSettings(actorSystem)
val (bindingFuture, closedFuture) = serverSource
  .mapAsync(serverSettings.getMaxConnections) { connection =>
    connection
      .flow
      .watchTermination()(Keep.right)
      .join(routes)
      .run()
      .recover {
        case NonFatal(_) => Done
      }
  }.toMat(Sink.ignore)(Keep.both)
  .run()

bindingFuture.map { serverBinding =>
  CoordinatedShutdown(system)
    .addTask(
      CoordinatedShutdown.PhaseServiceUnbind,
      "unbindHttp",
    ) { () =>
      serverBinding.unbind()
        .flatMap { _ => closedFuture }
    }
}
```

**Comment by dsilvasc** [comment]
junit repro test:

```scala
  // https://github.com/akka/akka/issues/16894#issuecomment-236729283
  @Test
  @SuppressWarnings(Array("org.wartremover.warts.While"))
  def testRepro(): Unit = for (n <- 1 to 4000) {
    println(s"run $n")
    import org.scalatest.junit.AssertionsForJUnit._
    import scala.concurrent.duration._
    implicit val actorSystem: ActorSystem = ActorSystem(name = "test")

    val done = new AtomicBoolean(false)
    val thread = new Thread(() => {
      // start the server, wait (run) until the actor system terminates, then stop

      import akka.http.scaladsl.Http
      import akka.http.scaladsl.model.HttpResponse
      import akka.http.scaladsl.server.{Route, RouteResult}
      import akka.http.scaladsl.server.RequestContext

      implicit val materializer: ActorMaterializer = ActorMaterializer(namePrefix = Some("test"))

      val route: Route = (context: RequestContext) => Future.successful(RouteResult.Complete(HttpResponse()))

      val bindingFuture = Http()
        .bindAndHandle(route, "0.0.0.0", port = 0)
      val httpBinding = Await.result(bindingFuture, Duration.Inf)
      println(s"server started at $httpBinding")

      def stopServer(): Unit = {
        import ExecutionContext.Implicits.global
        // import actorSystem.dispatcher

        println(s"stopping server")

        val terminateFuture =
          // Nothing to do when unbind was already called
          if (httpBinding.whenTerminationSignalIssued.isCompleted) Future.unit
          // Nothing to do when the actor system is already terminated
          else if (actorSystem.whenTerminated.isCompleted) Future.unit
          // Unbind the server port and wait 10 seconds for requests in flight to complete
          else httpBinding.terminate(10.seconds).map(_ => ())

        val stopFuture = terminateFuture.andThen {
          case _ => materializer.shutdown()
        }

        Await.result(stopFuture, Duration.Inf)
      }

      try Await.result(actorSystem.whenTerminated, Duration.Inf)
      catch {
        case _: InterruptedException =>
        // nothing to do when interrupted
      } finally stopServer()
      done.set(true)
      println(s"server thread done")
    })
    thread.start()

    // wait for the server to await termination
    while (thread.getState != Thread.State.WAITING) { // scalastyle:ignore
      Thread.sleep(5)
    }

    assert(!done.get)
    actorSystem.terminate()
    thread.join()
    assert(done.get)
  }
```

It happens fairly rarely, thus the lengthy loop.  Eventually this happens:

```
...
run 25
server started at ServerBinding(/0:0:0:0:0:0:0:0:50348)
stopping server
server thread done
run 26
Exception in thread "Thread-26" java.util.NoSuchElementException: Binding was unbound before it was completely finished
	at akka.stream.impl.io.ConnectionSourceStage$$anon$2.postStop(TcpStages.scala:161)
	at akka.stream.impl.fusing.GraphInterpreter.finalizeStage(GraphInterpreter.scala:571)
	at akka.stream.impl.fusing.GraphInterpreter.finish(GraphInterpreter.scala:313)
	at akka.stream.impl.fusing.GraphInterpreterShell.tryAbort(ActorGraphInterpreter.scala:626)
	at akka.stream.impl.fusing.ActorGraphInterpreter.$anonfun$postStop$1(ActorGraphInterpreter.scala:797)
	at akka.stream.impl.fusing.ActorGraphInterpreter.$anonfun$postStop$1$adapted(ActorGraphInterpreter.scala:797)
	at scala.collection.immutable.Set$Set1.foreach(Set.scala:95)
	at akka.stream.impl.fusing.ActorGraphInterpreter.postStop(ActorGraphInterpreter.scala:797)
	at akka.actor.Actor.aroundPostStop(Actor.scala:536)
	at akka.actor.Actor.aroundPostStop$(Actor.scala:536)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundPostStop(ActorGraphInterpreter.scala:670)
	at akka.actor.dungeon.FaultHandling.finishTerminate(FaultHandling.scala:210)
	at akka.actor.dungeon.FaultHandling.terminate(FaultHandling.scala:172)
	at akka.actor.dungeon.FaultHandling.terminate$(FaultHandling.scala:142)
	at akka.actor.ActorCell.terminate(ActorCell.scala:431)
	at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:529)
	at akka.actor.ActorCell.systemInvoke(ActorCell.scala:545)
	at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:283)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:235)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

org.scalatest.junit.JUnitTestFailedError: done.get() was false

	at org.scalatest.junit.AssertionsForJUnit.newAssertionFailedException(AssertionsForJUnit.scala:100)
	at org.scalatest.junit.AssertionsForJUnit.newAssertionFailedException$(AssertionsForJUnit.scala:99)
	at org.scalatest.junit.AssertionsForJUnit$.newAssertionFailedException(AssertionsForJUnit.scala:157)
	at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:501)
```
